116917: Fix connection lost edge case for locator

msdt93 opened the pull request:
- If it happened that connection is lost in the middle of reading the locator reply, it could happen that the request would get stuck until timeout expires.
 - Also, if it happens that recv call fails in less than 1 ms, considering the timeout had ms precision, the application could end up within an infinite loop.
 - So this covers the connection lost case and also the precision issue.
 - Also, added a new UT to verify the connection lost case.

Commit:
116917: Fix connection lost edge case for locator

 - If it happened that connection is lost in the middle of reading the locator
   reply, it could happen that the request would get stuck until timeout
   expires.
 - Also, if it happens that recv call fails in less than 1 ms,
   considering the timeout had ms precision, the application could end
   up within an infinite loop.
 - So this covers the connection lost case and also the precision issue.
 - Also, added a new UT to verify the connection lost case.

albertogpz commented on cppcache/src/StreamDataInput.cpp:
```diff
@@ -37,42 +37,51 @@ StreamDataInput::StreamDataInput(std::chrono::milliseconds timeout,
     bufferHead_ = streamBuf_.data();
     buffer_ = bufferHead_ + currentPosition;
-    bufferLength_ = newLength;
+    remainingTimeBeforeTimeout_ = remainingTime;
```
I suggest to move this line upwards (right after the calculation of `remainingTime` in line 50) and remove any other line in this method updating `remainingTimeBeforeTimeout_`.

Commit:
116917: Revision 1

 - Addressed comments from revision 1
 - Now remaining timeout is only updated once, making logic more robust.
